From 5ee2cc070ffe8c0852df5a42e538e0e34259f749 Mon Sep 17 00:00:00 2001
From: Rex Lu <rex.lu@mediatek.com>
Date: Tue, 11 Nov 2025 12:20:29 +0800
Subject: [PATCH 018/105] mtk: mt76: mt7996: check txfree done event on hwpath

align logan to check txfree done event dw1 bit15 when wed enabled.
to avoid driver read txfree done event before wed finish read.

---
 dma.c | 3 +++
 dma.h | 2 +-
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/dma.c b/dma.c
index 9245a3b5a..fca533467 100644
--- a/dma.c
+++ b/dma.c
@@ -592,6 +592,9 @@ mt76_dma_dequeue(struct mt76_dev *dev, struct mt76_queue *q, bool flush,
 			q->desc[idx].ctrl |= cpu_to_le32(MT_DMA_CTL_DMA_DONE);
 		else if (!(q->desc[idx].ctrl & cpu_to_le32(MT_DMA_CTL_DMA_DONE)))
 			return NULL;
+		else if (mt76_queue_is_wed_tx_free(q) &&
+			!(q->desc[idx].ctrl & cpu_to_le32(MT_DMA_CTL_M_DONE)))
+			return NULL;
 	}
 done:
 	q->tail = (q->tail + 1) % q->ndesc;
diff --git a/dma.h b/dma.h
index 4fd779e1a..1dc739ee4 100644
--- a/dma.h
+++ b/dma.h
@@ -11,7 +11,7 @@
 
 #define MT_DMA_CTL_SD_LEN1		GENMASK(13, 0)
 #define MT_DMA_CTL_LAST_SEC1		BIT(14)
-#define MT_DMA_CTL_BURST		BIT(15)
+#define MT_DMA_CTL_M_DONE		BIT(15)
 #define MT_DMA_CTL_SD_LEN0		GENMASK(29, 16)
 #define MT_DMA_CTL_LAST_SEC0		BIT(30)
 #define MT_DMA_CTL_DMA_DONE		BIT(31)
-- 
2.45.2

